File: Api/Application/TournamentService.cs
using Api.Domain;
using Api.Infrastructure;
using Api.Shared;

namespace Api.Application;

/// <summary>
/// Orchestrates Pokémon tournament workflow: data fetching, battle simulation, and result sorting.
/// </summary>
public class TournamentService : ITournamentService
{
    private readonly PokemonApiClient _pokemonApiClient;
    private readonly BattleSimulator _battleSimulator;

    public TournamentService(PokemonApiClient pokemonApiClient, BattleSimulator battleSimulator)
    {
        _pokemonApiClient = pokemonApiClient;
        _battleSimulator = battleSimulator;
    }

    /// <summary>
    /// Executes the full tournament process: fetches Pokémon, simulates battles,
    /// and returns sorted statistics based on the given criteria.
    /// </summary>
    /// <param name="sortBy">The field used for sorting (e.g., wins, losses, name, etc.).</param>
    /// <param name="sortDirection">The direction of sorting (asc or desc, default is asc).</param>
    /// <returns>A collection of Pokémon with calculated battle statistics.</returns>
    public async Task<IEnumerable<TournamentPokemon>> GetTournamentStatistics(string sortBy, string sortDirection = "asc")
    {
        var pokemons = await _pokemonApiClient.FetchRandomPokemonsAsync();
        
        foreach (var pokemon in pokemons)
        {
            pokemon.Name = StringHelper.Capitalize(pokemon.Name);
        }

        _battleSimulator.SimulateAllBattles(pokemons);
        return SortPokemons(pokemons, sortBy, sortDirection);
    }

    private IEnumerable<TournamentPokemon> SortPokemons(List<TournamentPokemon> pokemons, string sortBy, string direction)
    {
        if (string.IsNullOrWhiteSpace(sortBy))
            throw new ArgumentException("sortBy parameter cannot be null or empty", nameof(sortBy));

        Func<TournamentPokemon, object> keySelector = sortBy.ToLower() switch
        {
            "wins" => p => p.Wins,
            "losses" => p => p.Losses,
            "ties" => p => p.Ties,
            "name" => p => p.Name,
            "id" => p => p.Id,
            _ => throw new ArgumentException($"Invalid sortBy value: {sortBy}", nameof(sortBy))
        };

        var sorted = direction.Equals("asc", StringComparison.OrdinalIgnoreCase)
            ? pokemons.OrderBy(keySelector)
            : pokemons.OrderByDescending(keySelector);

        return sorted.ToList();
    }


}

File: frontend/src/app/services/tournament.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { TournamentPokemon, SortBy, SortDirection } from '../models';

@Injectable({
  providedIn: 'root'
})
export class TournamentService {
  private readonly apiUrl = 'http://localhost:5268/pokemon/tournament/statistics';

  constructor(private http: HttpClient) { }

  /**
   * Fetches tournament statistics from the API with sorting options
   * @param sortBy - Field to sort by (wins, losses, ties, name, id)
   * @param sortDirection - Sort direction (asc or desc)
   * @returns Observable of tournament pokemon array
   */
  getTournamentStatistics(
    sortBy: SortBy, 
    sortDirection: SortDirection = 'asc'
  ): Observable<TournamentPokemon[]> {
    const params = new HttpParams()
      .set('sortBy', sortBy)
      .set('sortDirection', sortDirection);

    return this.http.get<TournamentPokemon[]>(this.apiUrl, { params });
  }
}
